src/basic/unit-file.c

```diff
@@ -285,19 +285,49 @@ int unit_file_build_name_map(
                 }
 
                 FOREACH_DIRENT_ALL(de, d, log_warning_errno(errno, "Failed to read \"%s\", ignoring: %m", *dir)) {
-                        char *filename;
                         _unused_ _cleanup_free_ char *_filename_free = NULL;
                         _cleanup_free_ char *simplified = NULL;
-                        const char *suffix, *dst = NULL;
-                        bool valid_unit_name;
-
-                        valid_unit_name = unit_name_is_valid(de->d_name, UNIT_NAME_ANY);
+                        const char *dst = NULL;
+                        char *filename;
 
                         /* We only care about valid units and dirs with certain suffixes, let's ignore the
                          * rest. */
-                        if (!valid_unit_name &&
-                            !ENDSWITH_SET(de->d_name, ".wants", ".requires", ".d"))
-                                continue;
+
+                        if (unit_name_is_valid(de->d_name, UNIT_NAME_ANY)) {
+
+                                /* Accept a regular file or symlink whose name is a valid unit file name. */
+                                if (!IN_SET(de->d_type, DT_REG, DT_LNK))
+                                        continue;
```

Comment:
IIUC the code, then I'd like to invert the order of the conditions, and skip suffix check for directory when `paths` is not set:
```
if (IN_SET(de->d_type, DT_REG, DT_LNK)) {
    if (!unit_name_is_valid(...))
        continue;
} else if (de->d_type == DT_DIR) {
    if (!paths)
        continue;

    FOREACH_STRING(suffix, ".wants", ".requires", ".d") {
        ...;
    }
} else
    continue;
```

Reply:
i think the inversion makes sense, but i don#t grok what you mean by the `paths` check? we add everything to `paths` if it is set, so the check makes sense in any case?

Reply:
When `path_cache` is not set, then this does nothing for directories, So, it is ok to simply skip entries with `DT_DIR`. In other words, for the case `DT_DIR`, `filename` is allocated below, and soon freed. That's redundant.